File: app/src/main/java/com/mohsenoid/rickandmorty/data/network/NetworkClient.java
package com.mohsenoid.rickandmorty.data.network;

import com.mohsenoid.rickandmorty.data.network.dto.NetworkCharacterModel;
import com.mohsenoid.rickandmorty.data.network.dto.NetworkEpisodeModel;

import org.json.JSONException;

import java.io.IOException;
import java.util.List;

public interface NetworkClient {

    List<NetworkEpisodeModel> getEpisodes(int page) throws IOException, JSONException;

    List<NetworkCharacterModel> getCharactersByIds(List<Integer> characterIds) throws IOException, JSONException;

    NetworkCharacterModel getCharacterDetails(int characterId) throws IOException, JSONException;
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/data/network/NetworkClient.kt
package com.mohsenoid.rickandmorty.data.network

import com.mohsenoid.rickandmorty.data.network.dto.NetworkCharacterModel
import com.mohsenoid.rickandmorty.data.network.dto.NetworkEpisodesResponse
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface NetworkClient {

    @GET(value = NetworkConstants.EPISODE_ENDPOINT)
    suspend fun fetchEpisodes(
        @Query(value = NetworkConstants.PARAM_KEY_PAGE)
        page: Int
    ): Response<NetworkEpisodesResponse>

    @GET(value = "${NetworkConstants.CHARACTER_ENDPOINT}{${NetworkConstants.PATH_KEY_CHARACTER_IDS}}")
    suspend fun fetchCharactersByIds(
        @Path(value = NetworkConstants.PATH_KEY_CHARACTER_IDS)
        characterIds: List<Int>
    ): Response<List<NetworkCharacterModel>>

    @GET(value = "${NetworkConstants.CHARACTER_ENDPOINT}{${NetworkConstants.PATH_KEY_CHARACTER_ID}}")
    suspend fun fetchCharacterDetails(
        @Path(value = NetworkConstants.PATH_KEY_CHARACTER_ID)
        characterId: Int
    ): Response<NetworkCharacterModel>
}
